Enable Drag and Drop between SubViewports and Windows

Sauermann opened the pull request:
Make Drag and Drop an application-wide operation.
This allows to drop on Controls in other Viewports/Windows.

Also now all nodes in the SceneTree are notified about the Drag and Drop operation, with the exception of SubViewports that are not child-nodes of SubViewportContainers. This distinction seemed prudent, because these SubViewports need to be displayed manually and this requires manual setup also for Drag and Drop.

Unit-tests for Drag and Drop in different Windows & SubViewports are also added.

In order to achieve this, the following changes were made:
- `Viewport::_update_mouse_over` from #67791 is used to identify the drop-target (even within nested viewports)
- This Control is used as a basis for the Drop-operation, which replaces the previous algorithm, that was only aware of topmost Viewports.
- In order to handle DnD within non-default SubViewports (see for an example https://github.com/godotengine/godot-demo-projects/pull/925), it becomes necessary to track drag-operations in each (for the lack of a better name) section in the scene-tree.
- `is_directly_attached_to_screen` is replaced by `get_section_rootviewport`. which is of the same complexity, but additionally returns the topmost Viewport within the section
- `_propagate_viewport_notification` had to be adjusted to account for sections
- `DisplayServerX11::delete_sub_window` needed to be adjusted similar to #80142, since the event-callback is also needed on linuxbsd during Window-deletion

Relations:
- alternative to #71509 and has a broader scope
- fix #28522
- fix #44390
- fix #59622
- fix #77665
- fix no longer (will be done separately) #67186 
- fix duplicate #62384
- fix duplicate #68063
- was previously blocked by #71972
- utilizes #67791
- follow-up to #80142
- related to #64949
- related to #20881

MRP: [BugDragNDrop.zip](https://github.com/godotengine/godot/files/10468628/BugDragNDrop.zip) (updated 2023-01-20)

The Drag-Preview in the following Video is outdated, because that will be done in a different PR.

https://user-images.githubusercontent.com/6299227/196206434-2b8cba24-5993-40d2-a8dc-22749b7ccd27.mp4

Updated 2023-01-28: Fix merge conflict
Updated 2023-03-20: Fix merge conflict
Updated 2023-05-30: Linked another bug-report, that this PR fixes. Resolve problematic testcases. Fix merge conflict.
Updated 2023-06-07: Fix merge conflict
Updated 2023-06-08: Fix issue with docks
Updated 2023-06-19: Fix issue with simultaneous drag starts
Updated 2023-07-06: Fix issue with Gui in 3D Demo, Add Unit-tests
Updated 2023-07-19: Fix merge conflict
Updated 2023-08-02: Fix merge conflict with #67791 and #79805
Updated 2023-08-12: Fix merge conflict
Updated 2023-09-18: Fix merge conflict with #81551
Updated 2024-01-03: Fix merge conflict with #86511 and verified that #86493 doesn't get reintroduced.
Updated 2024-05-03: Rebased after #91425 and verified that #91387 doesn't get reintroduced.
Updated 2024-09-15: Rebased after 4.3-release for good measure and verified the functionality in the MRP.

rokiyo commented:
Fixes #67186
Potentially supersedes #64949 ?

Sauermann commented:
@rokiyo Thanks, I did not see these.

Sauermann commented on scene/main/viewport.cpp:
```diff
@@ -1797,104 +1802,155 @@ void Viewport::_gui_input_event(Ref<InputEvent> p_event) {
+				if (previous_drop_target && previous_drop_target->get_viewport()->get_base_window() != start_base_window &&
+						(start_base_window == this || is_ancestor_of(previous_drop_target->get_viewport()->get_base_window()))) {
+					_gui_call_notification(previous_drop_target, Control::NOTIFICATION_MOUSE_EXIT);
+				}
```
This became rather complicated because depending on drag source and target the code is executed in one or more different viewports.
While this current solution works in a large amount of situations, a better solution would be to execute drop-handling only in a single viewport.
For the moment this works for all my testcases, but I intend to revisit this after #58334, since that other PR will change the order in which viewports handle events and so a solution would require adjustment anyway.

Sauermann replied:
> [...] a better solution would be to execute drop-handling only in a single viewport

I just verified, that this is not possible, because `SubViewports` can embed `Windows` and that combination makes it necessary to execute this code also in nested `Viewports`.

bruvzg reviewed (approved):
Seems to be working fine.

Note: At least on macOS, mouse cursor is changed to `<->` when crossing resizable window border, and not restored to DnD cursor afterwards (until drop target is hovered).

Sauermann commented on scene/main/viewport.cpp:
```diff
@@ -1563,43 +1567,43 @@ void Viewport::_gui_input_event(Ref<InputEvent> p_event) {
 			}
 
-			if (gui.drag_data.get_type() != Variant::NIL && mb->get_button_index() == MouseButton::LEFT) {
+			if (get_tree()->get_root()->gui.drag_data.get_type() != Variant::NIL && mb->get_button_index() == MouseButton::LEFT) {
```
With this PR, `gui.drag_data` is only used on the root Viewport. I am thinking about making it a Viewport static class-variable. This approach would also help with readability, since `get_tree()->get_root()` is no longer needed.

Sauermann commented:
> Note: At least on macOS, mouse cursor is changed to <-> when crossing resizable window border, and not restored to DnD cursor afterwards (until drop target is hovered).

That sounds like Godot believes that the mouse cursor is still `FORBIDDEN` while macOS changed it to `RESIZE`.
I believe that this is a platform-specific problem, that should be fixed in the macOS-display server?

Sauermann commented on scene/main/viewport.cpp:
```diff
@@ -1797,104 +1802,155 @@ void Viewport::_gui_input_event(Ref<InputEvent> p_event) {
-						swrect.size.y += title_height;
+				if (embedder_window) {
+					viewport_pos = (embedder_window->get_final_transform().affine_inverse() * embedder_window->_get_input_pre_xform()).xform(screen_mouse_pos - embedder_window->get_position());
+					if (drag_preview->get_viewport() != embedder_window) {
```
Causes nullptr-exception when no Preview is provided.

Sauermann commented:
This implementation could be greatly simplified, if the functionality of #67791 was available.
Converting to draft, until a decision on #67791 is made.

Chaosus commented:
@Sauermann how is your progress with this PR?

Sauermann commented:
@Chaosus 
This PR is blocked by progress on #67791.
Both implement the same algorithm in different locations and #67791 implements it in a location, where it additionally fixes several bugs (which can not be fixed by this PR). So my recommendation would be to review and merge #67791 before I continue working on this PR.

Also #58334 will cause a few small adjustments to this PR, but I consider them minor. Ideally I would prefer, that #58334 should be reviewed before this one, but that is not strictly necessary.

reduz commented:
I wrote #71509 as a far simpler way to solve this. I believe It fixes drags/drops accross viewports more elegantly. What is missing is proper drag previews across windows, but the correct solution to this in my mind is to use borderless, transparent, passthrough windows similar to what we use with tooltips.

Sauermann commented:
The complexity in this PR comes from the fact that is solves three problems:
1. Drag and Drop in different Windows/SubViewports
2. Drag-Preview in different Viewports (The idea of using borderless, transparent, passthrough windows is superior to my own current approach)
3. Fix mouse-enter-exit-signals during Drag and Drop

In order to be able to compare it with #71509 (which aims to solve 1), I believe, that I should extract the part of this PR, that solves 1, so that they can be compared better. It looks like I am struggling again with packing too much content into a single PR.

From a quick comparison, this approach makes Drag and Drop an application-wide operation, while #71509 uses a heuristic to determine a Viewport to use as a base for the Drag and Drop operation.

Sauermann committed:
Enable Drag and Drop for SubViewports and Windows

Make Drag and Drop an application-wide operation.
This allows do drop on Controls in other Viewports/Windows.

In order to achieve this, `Viewport::_update_mouse_over` is adjusted to
remember the Control, that the mouse is over (possibly within nested
viewports). This Control is used as a basis for the Drop-operation, which
replaces the previous algorithm, which was only aware of the topmost
Viewport.

Also now all nodes in the SceneTree are notified about the Drag and Drop
operation, with the exception of SubViewports that are not children of
SubViewportContainers.

Sauermann commented:
I have updated this PR based on the availability of the required PR #67791 and #58334.

Sauermann commented:
With #67791 merged, now this PR is ready.
I have tested each of the linked issues again on X11 and on Windows and made necessary adjustments to account for other changes in the code-base that influenced this PR since its initial conception.

KoBeWi commented:
Drag preview between Window viewports is planned for another PR?

KoBeWi commented on scene/main/viewport.h:
```diff
@@ -360,24 +360,25 @@ class Viewport : public Node {
+		bool dragging = false; // Is true in the viewport, in which dragging started, while dragging is active.
+		bool global_dragging = false; // Is true while dragging is active. Only used in Root-Viewport and SubViewport not in SubViewpotContainer.
 		bool drag_successful = false;
+		Control *target_control; // Control that the mouse is over in the innermost nested Viewport. Only used in Root-Viewport and SubViewport not in SubViewpotContainer.
```
```suggestion
		Control *target_control = nullptr; // Control that the mouse is over in the innermost nested Viewport. Only used in Root-Viewport and SubViewport not in SubViewpotContainer.
```

Sauermann replied:
Makes sense, thanks.

KoBeWi commented on scene/main/viewport.h:
```diff
@@ -626,6 +627,7 @@ class Viewport : public Node {
 	bool is_handling_input_locally() const;
 
 	bool gui_is_dragging() const;
+	bool gui_is_globally_dragging() const;
```
`gui_is_dragging()` is exposed, while `gui_is_globally_dragging()` is not.
I wonder if the global drag couldn't replace the former one. Is there any use-case for the old method? Seems like it's just less reliable.

Sauermann replied:
Previously `gui_is_dragging()` could be used, to query, if a dragging action is active in the current Viewport.

With the changes of this PR drag-and-drop becomes an operation that spans multiple Viewports, so now there are two concepts:
1. Is drag-and-drop active, that started in this Viewport? (This PRs `gui_is_dragging()`)
2. Is drag-and-drop active, that can drop in the current Viewport? (This PRs `gui_is_globally_dragging()`)

For the setting, that there is only a single Viewport in the scene, both behave identical.

After giving it more thought, I believe the second concept is more important for users, So I will update the PR and simply replace `gui_is_dragging()` by the second concept.

KoBeWi commented on scene/main/viewport.h:
```diff
@@ -807,7 +809,7 @@ class SubViewport : public Viewport {
 	virtual Transform2D get_screen_transform_internal(bool p_absolute_position = false) const override;
 	virtual Transform2D get_popup_base_transform() const override;
-	virtual bool is_directly_attached_to_screen() const override;
+	virtual Viewport *get_section_rootviewport() const override;
```
```suggestion
	virtual Viewport *get_section_root_viewport() const override;
```
`rootviewport()` is some odd spelling. Any specific reason why was it used?

Sauermann replied:
No specific reason. Will change it according to your suggestion.

akien-mga commented:
Could use testing on macOS (and review if you can) @bruvzg

EiTaNBaRiBoA commented:
This is if not the most important feature to distinguish between flutter and godot.
If godot will support it, it literally ends flutter, wpf, Qt and many more. I work as a software engineer and multi window communication is a deal breaker for many companies. This feature solves it and bring plenty of possibilities in the future. Please add this

akien-mga commented on doc/classes/Viewport.xml:
```diff
@@ -144,7 +144,7 @@
 			<return type="bool" />
 			<description>
-				Returns [code]true[/code] if the viewport is currently performing a drag operation.
+				Returns [code]true[/code] if a drag operation is currently performing, that can drop in this viewport.
```
```suggestion
				Returns [code]true[/code] if a drag operation is currently ongoing and where the drop action could happen in this viewport.
```

Sauermann replied:
Thanks for the code-style and the comment corrections. I have included them.

akien-mga commented on scene/main/viewport.cpp:
```diff
@@ -2006,105 +2011,29 @@ void Viewport::_gui_input_event(Ref<InputEvent> p_event) {
 
 		if (gui.dragging) {
-			// Handle drag & drop.
+			// Handle drag & drop. This happens in the viewport, where dragging started.
```
```suggestion
			// Handle drag & drop. This happens in the viewport where dragging started.
```

akien-mga commented on scene/main/viewport.cpp:
```diff
@@ -2006,105 +2011,29 @@ void Viewport::_gui_input_event(Ref<InputEvent> p_event) {
 
-		if (DisplayServer::get_singleton()->has_feature(DisplayServer::FEATURE_CURSOR_SHAPE) && !Object::cast_to<SubViewportContainer>(over)) {
+		if (DisplayServer::get_singleton()->has_feature(DisplayServer::FEATURE_CURSOR_SHAPE) && (gui.dragging || (!section_root->gui.global_dragging && !Object::cast_to<SubViewportContainer>(over)))) {
+			// If dragging is active, then set cursor shape only from Viewport where dragging started.
+			// If dragging is inactive, then set cursor shape only when not over a SubViewportContainer.
```
```suggestion
			// If dragging is active, then set the cursor shape only from the Viewport where dragging started.
			// If dragging is inactive, then set the cursor shape only when not over a SubViewportContainer.
```

akien-mga commented on scene/main/viewport.h:
```diff
@@ -375,16 +374,18 @@ class Viewport : public Node {
 		Point2 drag_accum;
 		bool drag_attempted = false;
-		Variant drag_data;
+		Variant drag_data; // Only used in Root-Viewport and SubViewports, that are not children of a SubViewpotContainer.
```
```suggestion
		Variant drag_data; // Only used in root Viewport and SubViewports, that are not children of a SubViewportContainer.
```

akien-mga commented on scene/main/viewport.h:
```diff
@@ -375,16 +374,18 @@ class Viewport : public Node {
 		HashSet<ObjectID> canvas_parents_with_dirty_order;
 		int canvas_sort_index = 0; //for sorting items with canvas as root
-		bool dragging = false;
+		bool dragging = false; // Is true in the viewport, in which dragging started, while dragging is active.
+		bool global_dragging = false; // Is true while dragging is active. Only used in Root-Viewport and SubViewports, that are not children of a SubViewpotContainer.
 		bool drag_successful = false;
+		Control *target_control = nullptr; // Control that the mouse is over in the innermost nested Viewport. Only used in Root-Viewport and SubViewports, that are not children of a SubViewpotContainer.
```
```suggestion
		bool dragging = false; // Is true in the viewport in which dragging started while dragging is active.
		bool global_dragging = false; // Is true while dragging is active. Only used in root Viewport and SubViewports that are not children of a SubViewportContainer.
		bool drag_successful = false;
		Control *target_control = nullptr; // Control that the mouse is over in the innermost nested Viewport. Only used in root Viewport and SubViewports that are not children of a SubViewportContainer.
```

akien-mga commented on tests/scene/test_viewport.h:
```diff
@@ -1107,12 +1123,10 @@ TEST_CASE("[SceneTree][Viewport] Controls and InputEvent handling") {
-		// FIXME: Testing Drag and Drop with non-embedded windows would require DisplayServerMock additions
-		// FIXME: Drag and Drop currently doesn't work with embedded Windows and SubViewports - not testing.
-		// See https://github.com/godotengine/godot/issues/28522 for example.
+		// Note: Testing Drag and Drop with non-embedded windows would require DisplayServerMock additions
```
```suggestion
		// Note: Testing Drag and Drop with non-embedded windows would require DisplayServerMock additions.
```

akien-mga reviewed (approved):
Looks good to me. Let's merge if you think now would be a good time.

akien-mga commented:
We decided to keep this for the 4.4 cycle, as it's pretty late now in the 4.3 cycle and this PR changes behavior in a way that may clash with user expectations. So we don't want to rush it / have to reconsider the design during the beta/RC stage.

akien-mga commented:
So now would be a good time to merge this for 4.4 I think?

akien-mga commented:
Thanks!

DS204ever commented:
Hello @Sauermann, by any chance do you still have the fix for the drag preview that you mentioned? I didn't find any open PR that solves this issue.

rokiyo commented:
> Hello @Sauermann, by any chance do you still have the fix for the drag preview that you showcased in the video? I didn't find any open PR that solves this issue.

In a similar vein, is the issue in the video the same as the one mentioned in #67186 ? I can't quite tell if they're the same, but that issue seems to be the only one still open?

DS204ever commented:
Yes I think it's the same issue. I have been messing around with the problem for a few hours now and I am close to getting something working, I will keep updating this topic.

Sauermann commented:
> Hello @Sauermann, by any chance do you still have the fix for the drag preview that you showcased in the video? I didn't find any open PR that solves this issue.

@DS204ever No I don't have the code anymore, because we discarded my approach.
